[CALCITE-3424] AssertionError thrown for user-defined table function with array argument

ihuzenko opened the pull request.

danny0405 commented on core/src/test/java/org/apache/calcite/util/Smalls.java:
```diff
@@ -182,6 +184,10 @@ public void close() {
   }
 
+  public static QueryableTable generateStrings2(final List<Integer> list) {
+    return generateStrings(list.size());
```
Can we give it a more meaning name ? not XXX2 ?

ihuzenko replied:
good catch, done.

danny0405 commented on core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java:
```diff
@@ -251,13 +253,25 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
+  private static boolean isCollection(Class javaClass) {
+    return javaClass == Collection.class
+        || Arrays.stream(javaClass.getInterfaces()).anyMatch(JavaTypeFactoryImpl::isCollection);
+  }
```
This would form a n * n *.. recursion, i don't think we need that, for each super class, only `javaClass == Collection.class` is enough.

danny0405 commented on core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java:
```diff
@@ -251,13 +253,25 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
+        nonNullType = typeFactory.createArrayType(typeFactory.createSqlType(SqlTypeName.ANY), -1);
+      } else {
+        nonNullType = typeFactory.createSqlType(type.getSqlTypeName());
+      }
```
Can you reference `JavaToSqlTypeConversionRules` for the type mapping ?

ihuzenko replied:
There is no need to use ```JavaToSqlTypeConversionRules``` here because ```type``` contains SqlTypeName. Most probably this check just need to be rewritten to switch case based on the enum values and further java class investigation isn't necessary here. To be honest the API of ```typeFactory``` which promises to ```createSqlType(...)``` for sqlTypeName and implicitly throws assertion errors is really bad.

danny0405 replied:
I didn't mean to use `JavaToSqlTypeConversionRules` directly, i meant to match all the java collection classes that synced with what `JavaToSqlTypeConversionRules` has already matched.

yanlin-Lynn commented on core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java:
```diff
@@ -251,13 +253,25 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
-          type.isNullable());
+      Class javaClass = ((JavaType) type).getJavaClass();
+      final RelDataType nonNullType;
+      if (javaClass.isArray()) {
```
what about map type?

ihuzenko replied:
Issue described in Jira ticket relates to ```SqlTypeFactoryImpl```'s assertBasic method: 
```java
  private void assertBasic(SqlTypeName typeName) {
    assert typeName != null;
    assert typeName != SqlTypeName.MULTISET
        : "use createMultisetType() instead";
    assert typeName != SqlTypeName.ARRAY
        : "use createArrayType() instead";
    assert typeName != SqlTypeName.ROW
        : "use createStructType() instead";
    assert !SqlTypeName.INTERVAL_TYPES.contains(typeName)
        : "use createSqlIntervalType() instead";
  }
```
For map type assertion isn't fired here and I think it's rare case that someone will accept map arguments (or others from the method body) into custom table function.

yanlin-Lynn replied:
I'm not sure that's a bug or is just designed.
I create a PR(https://github.com/apache/calcite/pull/1521) trying to fix it

yanlin-Lynn commented on core/src/test/java/org/apache/calcite/test/TableFunctionTest.java:
```diff
@@ -105,6 +105,24 @@
     }
   }
 
+  @Test public void testTableFunctionWithArrayParameter() throws SQLException {
```
may be add a case for map type parameter?

danny0405 replied:
I think array is enough, just like what the commit describes.

jinxing64 commented on core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java:
```diff
@@ -251,9 +251,16 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
+      final RelDataType relDataType;
+      if (sqlTypeName == SqlTypeName.ARRAY) {
+        RelDataType elementType = type.getComponentType() == null
+            ? typeFactory.createSqlType(SqlTypeName.ANY) : type.getComponentType();
```
Does it need special consideration when it's a nested array, i.e. the element type of array is also of array

ihuzenko replied:
Could you please give an example of table function for which it's necessary to accept nested array as argument ?

danny0405 replied:
You can use `SqlTypeUtil#isArray` to decide if a type is `ARRAY` type.

ihuzenko commented:
Hi @danny0405 , I've updated JavaTypeFactoryImpl according to your suggestion, rebased on latest master and squashed commits. Please take a look again.
